server/src/main/java/org/elasticsearch/action/bulk/BulkShardRequestInferenceProvider.java

```diff
@@ -0,0 +1,296 @@
+        };
+        try (var bulkItemReqRef = new RefCountingRunnable(onInferenceComplete)) {
+            BulkItemRequest[] items = bulkShardRequest.items();
+            for (int i = 0; i < items.length; i++) {
+                BulkItemRequest bulkItemRequest = items[i];
+                // Bulk item might be null because of previous errors, skip in that case
+                if (bulkItemRequest != null) {
+                    performInferenceOnBulkItemRequest(
+                        bulkItemRequest,
+                        fieldsForModels,
+                        i,
+                        onBulkItemFailure,
+                        failedItems,
+                        bulkItemReqRef.acquire()
+                    );
+                }
+            }
+        }
+    }
+
+    private void performInferenceOnBulkItemRequest(
+        BulkItemRequest bulkItemRequest,
+        Map<String, Set<String>> fieldsForModels,
+        Integer itemIndex,
+        BiConsumer<BulkItemRequest, Exception> onBulkItemFailure,
+        Set<Integer> failedItems,
+        Releasable releaseOnFinish
+    ) {
+
+        DocWriteRequest<?> docWriteRequest = bulkItemRequest.request();
+        Map<String, Object> sourceMap = null;
+        if (docWriteRequest instanceof IndexRequest indexRequest) {
+            sourceMap = indexRequest.sourceAsMap();
+        } else if (docWriteRequest instanceof UpdateRequest updateRequest) {
+            sourceMap = updateRequest.docAsUpsert() ? updateRequest.upsertRequest().sourceAsMap() : updateRequest.doc().sourceAsMap();
+        }
+        if (sourceMap == null || sourceMap.isEmpty()) {
+            releaseOnFinish.close();
+            return;
+        }
+        final Map<String, Object> docMap = new ConcurrentHashMap<>(sourceMap);
+
+        // When a document completes processing, update the source with the inference
+        try (var docRef = new RefCountingRunnable(() -> {
+            if (docWriteRequest instanceof IndexRequest indexRequest) {
+                indexRequest.source(docMap);
+            } else if (docWriteRequest instanceof UpdateRequest updateRequest) {
+                if (updateRequest.docAsUpsert()) {
+                    updateRequest.upsertRequest().source(docMap);
+                } else {
+                    updateRequest.doc().source(docMap);
+                }
+            }
+            releaseOnFinish.close();
+        })) {
+
+            for (Map.Entry<String, Set<String>> fieldModelsEntrySet : fieldsForModels.entrySet()) {
+                String modelId = fieldModelsEntrySet.getKey();
+
+                @SuppressWarnings("unchecked")
+                Map<String, Object> rootInferenceFieldMap = (Map<String, Object>) docMap.computeIfAbsent(
+                    ROOT_INFERENCE_FIELD,
+                    k -> new HashMap<String, Object>()
+                );
+
+                List<String> inferenceFieldNames = getFieldNamesForInference(fieldModelsEntrySet, docMap);
```

Comment:
We could simplify the method signature to `getFieldNamesForInference(Set<String> inferenceFields, Map<String, Object> docMap)` by passing `fieldModelsEntrySet.getValue()`. This would also make it clearer that this helper method doesn't use the model ID.